File: apps/Model Plane/go/pkg/jetstream/jetstream.go
// Package jetstream provides declarative, language-agnostic specifications
// for NATS JetStream streams and consumers that back the Model Plane event
// bus. The JSON shape mirrors the Rust mp-events crate byte-for-byte so the
// same fixtures validate on both sides.
package jetstream

import (
	"encoding/json"
	"errors"
)

// Retention policy values.
const (
	RetentionLimits    = "limits"
	RetentionInterest  = "interest"
	RetentionWorkQueue = "workqueue"
)

// Storage backend values.
const (
	StorageFile   = "file"
	StorageMemory = "memory"
)

// Discard policy values.
const (
	DiscardOld = "old"
	DiscardNew = "new"
)

// Ack policy values.
const (
	AckExplicit = "explicit"
	AckAll      = "all"
	AckNone     = "none"
)

// StreamSpec is the declarative specification of a JetStream stream.
type StreamSpec struct {
	Name       string   `json:"name"`
	Subjects   []string `json:"subjects"`
	Retention  string   `json:"retention"`
	Storage    string   `json:"storage"`
	MaxAgeSecs uint64   `json:"max_age_secs"`
	Replicas   uint8    `json:"replicas"`
	Discard    string   `json:"discard"`
}

// Validate returns an error if any required field is missing or zero.
func (s *StreamSpec) Validate() error {
	if s.Name == "" {
		return errors.New("missing required field: name")
	}
	if len(s.Subjects) == 0 {
		return errors.New("missing required field: subjects")
	}
	for _, subj := range s.Subjects {
		if subj == "" {
			return errors.New("missing required field: subjects")
		}
	}
	if s.MaxAgeSecs == 0 {
		return errors.New("missing required field: max_age_secs")
	}
	if s.Replicas == 0 {
		return errors.New("missing required field: replicas")
	}
	return nil
}

// Encode serializes the spec to JSON bytes.
func (s *StreamSpec) Encode() ([]byte, error) {
	return json.Marshal(s)
}

// DecodeStream deserializes a StreamSpec from JSON bytes.
func DecodeStream(data []byte) (*StreamSpec, error) {
	var s StreamSpec
	if err := json.Unmarshal(data, &s); err != nil {
		return nil, err
	}
	return &s, nil
}

// ConsumerSpec is the declarative specification of a JetStream consumer.
type ConsumerSpec struct {
	Durable       string `json:"durable"`
	AckPolicy     string `json:"ack_policy"`
	AckWaitSecs   uint64 `json:"ack_wait_secs"`
	MaxDeliver    uint32 `json:"max_deliver"`
	FilterSubject string `json:"filter_subject"`
}

// Validate returns an error if any required field is missing or zero.
func (c *ConsumerSpec) Validate() error {
	if c.Durable == "" {
		return errors.New("missing required field: durable")
	}
	if c.AckWaitSecs == 0 {
		return errors.New("missing required field: ack_wait_secs")
	}
	if c.MaxDeliver == 0 {
		return errors.New("missing required field: max_deliver")
	}
	if c.FilterSubject == "" {
		return errors.New("missing required field: filter_subject")
	}
	return nil
}

// Encode serializes the spec to JSON bytes.
func (c *ConsumerSpec) Encode() ([]byte, error) {
	return json.Marshal(c)
}

// DecodeConsumer deserializes a ConsumerSpec from JSON bytes.
func DecodeConsumer(data []byte) (*ConsumerSpec, error) {
	var c ConsumerSpec
	if err := json.Unmarshal(data, &c); err != nil {
		return nil, err
	}
	return &c, nil
}

File: apps/Model Plane/rust/crates/mp-events/src/jetstream.rs
//! JetStream stream-as-code and consumer-as-code specifications.
//!
//! These types provide a declarative, language-agnostic description of the
//! NATS JetStream streams and consumers that back the Model Plane event bus.
//! The same JSON fixtures are consumed by the Go parity tests to keep the
//! Rust and Go implementations aligned byte-for-byte.

use serde::{Deserialize, Serialize};

/// Retention policy for a JetStream stream.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RetentionPolicy {
    /// Retain messages until limits are hit.
    Limits,
    /// Retain messages while interest (consumers) exists.
    Interest,
    /// Work-queue semantics; messages removed after ack.
    WorkQueue,
}

/// Storage backend for a JetStream stream.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum StorageType {
    /// File-backed storage.
    File,
    /// In-memory storage.
    Memory,
}

/// Discard policy when stream limits are hit.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DiscardPolicy {
    /// Discard oldest messages first.
    Old,
    /// Reject new messages.
    New,
}

/// Acknowledgement policy for a JetStream consumer.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AckPolicy {
    /// Each message must be explicitly acked.
    Explicit,
    /// Ack of a sequence acks all prior messages.
    All,
    /// No acks required.
    None,
}

/// Declarative specification of a JetStream stream.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct StreamSpec {
    /// Stream name (must be unique within the JetStream account).
    pub name: String,
    /// Subjects captured by this stream.
    pub subjects: Vec<String>,
    /// Retention policy.
    pub retention: RetentionPolicy,
    /// Storage backend.
    pub storage: StorageType,
    /// Maximum message age in seconds.
    pub max_age_secs: u64,
    /// Number of replicas (>= 1).
    pub replicas: u8,
    /// Discard policy when limits are hit.
    pub discard: DiscardPolicy,
}

impl StreamSpec {
    /// Validate that all required fields are present and well-formed.
    ///
    /// # Errors
    ///
    /// Returns [`JetStreamSpecError::MissingField`] if a required field is empty or zero.
    pub fn validate(&self) -> Result<(), JetStreamSpecError> {
        if self.name.is_empty() {
            return Err(JetStreamSpecError::MissingField("name"));
        }
        if self.subjects.is_empty() {
            return Err(JetStreamSpecError::MissingField("subjects"));
        }
        if self.subjects.iter().any(std::string::String::is_empty) {
            return Err(JetStreamSpecError::MissingField("subjects"));
        }
        if self.max_age_secs == 0 {
            return Err(JetStreamSpecError::MissingField("max_age_secs"));
        }
        if self.replicas == 0 {
            return Err(JetStreamSpecError::MissingField("replicas"));
        }
        Ok(())
    }

    /// Encode this spec to JSON bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if serialization fails.
    pub fn to_json_bytes(&self) -> Result<Vec<u8>, serde_json::Error> {
        serde_json::to_vec(self)
    }

    /// Decode a spec from JSON bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if deserialization fails.
    pub fn from_json_bytes(bytes: &[u8]) -> Result<Self, serde_json::Error> {
        serde_json::from_slice(bytes)
    }
}

/// Declarative specification of a JetStream consumer.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ConsumerSpec {
    /// Durable consumer name.
    pub durable: String,
    /// Ack policy.
    pub ack_policy: AckPolicy,
    /// Acknowledgement wait in seconds.
    pub ack_wait_secs: u64,
    /// Maximum redelivery attempts.
    pub max_deliver: u32,
    /// Filter subject for this consumer.
    pub filter_subject: String,
}

impl ConsumerSpec {
    /// Validate that all required fields are present and well-formed.
    ///
    /// # Errors
    ///
    /// Returns [`JetStreamSpecError::MissingField`] if a required field is empty or zero.
    pub fn validate(&self) -> Result<(), JetStreamSpecError> {
        if self.durable.is_empty() {
            return Err(JetStreamSpecError::MissingField("durable"));
        }
        if self.ack_wait_secs == 0 {
            return Err(JetStreamSpecError::MissingField("ack_wait_secs"));
        }
        if self.max_deliver == 0 {
            return Err(JetStreamSpecError::MissingField("max_deliver"));
        }
        if self.filter_subject.is_empty() {
            return Err(JetStreamSpecError::MissingField("filter_subject"));
        }
        Ok(())
    }

    /// Encode this spec to JSON bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if serialization fails.
    pub fn to_json_bytes(&self) -> Result<Vec<u8>, serde_json::Error> {
        serde_json::to_vec(self)
    }

    /// Decode a spec from JSON bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if deserialization fails.
    pub fn from_json_bytes(bytes: &[u8]) -> Result<Self, serde_json::Error> {
        serde_json::from_slice(bytes)
    }
}

/// Errors from JetStream spec operations.
#[derive(Debug, thiserror::Error)]
pub enum JetStreamSpecError {
    #[error("missing required field: {0}")]
    MissingField(&'static str),
}
